### ISSUE TITLE
How to know the max width of file?

### ISSUE THREAD

**Issue by acheronfail** [google/xi-editor#479]
This issue will be largely dependent on how xi-core's internal measurement code works (character widths for unicode, non-monospace fonts, etc) but I've run into this issue while developing a frontend for xi-editor.

Is there a way for the frontend to know the character length of the longest line?

If the first 100 or so lines are short, but there may be a long line further down in the document. The issue is that the frontend doesn't know this, and won't be able to horizontally scroll accordingly until it encounters the longest line.

**Comment by cmyr** [comment]
When we've been talking about width measurement to date we've been mostly focused on the general case, accounting for unicode and non-monospace typefaces. The best issue related to this is here: https://github.com/google/xi-mac/issues/4 but there should be a xi-core issue as well.

I know we've discussed this particular request as well at some point, but I don't remember the conclusion.

One thing to bear in mind is that even using monospace ASCII, 'character count' is not representative of width; tabs are the most obvious problem. Once you bring in unicode, this gets, uh, truer.

From a more UX pov, I personally dislike when there are horizontal scrollers when now scrollable content is visible; but I do think there should be room for flexibility for frontend authors.

I'd be curious to hear @raphlinus's thoughts. If we were going to do this the reasonable place would be the `update` RPC, which could include a `max_line_len` field whenever that value changes.

**Comment by acheronfail** [comment]
Yeah non-monospace fonts and unicode are definitely a problem with this. I've begun handling non-monospace fonts and unicode in my front end ([xi-electron](https://github.com/acheronfail/xi-electron)). What I've got so far works, but will definitely choke on longer lines.

I'm wary of over-implementing support for this and creating more efficient data structures at the moment because I'm unsure of the current direction of xi-editor in this regard - so my frontend manually measures the fonts/unicode/etc in order to correctly draw selections/cursor positions, etc.

It is definitely a curious problem when we think about the architecture of xi-editor - having a disconnected backend and frontend - I'm not sure what we can do about it, my ideas are:

* send characters widths to the core from the frontend (and have the core take them into account)
* leave it solely up to the frontend (but this affects line wrapping handling and other core functionality)
* anything else?

I'm keen to hear some of your thoughts on the potential solutions to this problem 😄

**Comment by cmyr** [comment]
Yea, I can see an argument for `max_line_len`. Happy to implement, but would want @raphlinus to chime in first.

For the second bit: last I heard we're leaning towards (approximately) the first option; core will request from the client the measurements for various glyph runs (for a given set of font attributes), which will be cached in core.
